### ISSUE TITLE
libtorch Segmentation Fault: RHEL 7 - easy to reproduce

### ISSUE THREAD

**Issue by sflc6** [pytorch/pytorch#19969]
## 🐛 Bug

tl; dr; I traced a (non-pretrained) resnet18 model in python and saved this to a .pt. I successfully loaded this .pt in a C++ program. Model inference seems to work. When the C++ program exits, I encounter a segmentation fault.

## To Reproduce

Here are code snippets to minimally reproduce the issue:

(1) First, generate the model by running the following python script. If successful, you should see a 0 print out. I'm able to run this code with no problems.

```import torch
import torchvision

# An instance of your model.
model = torchvision.models.resnet18()

# An example input you would normally provide to your model's forward() method.
example = torch.rand(1, 3, 224, 224).cuda()

# Use torch.jit.trace to generate a torch.jit.ScriptModule via tracing.
model.cuda()

before = model(example)
print(before.shape)

traced_script_module = torch.jit.trace(model, example)
traced_script_module.save('hard.pt')

load_module = torch.jit.load("hard.pt")
after = load_module(example)

print(torch.sum(before - after))
```

(2) Here is the C++ program I run that generates a segmentation fault. To run, do `./example-app /path/to/hard.pt`

```
#define _GLIBCXX_USE_CXX11_ABI 0
#include <torch/script.h> // One-stop header.

#include <iostream>
#include <memory>
#include <vector>

int main(int argc, const char* argv[]) {
  if (argc != 2) {
    std::cerr << "usage: example-app <path-to-exported-script-module>\n";
    return -1;
  }

  torch::Tensor tensor = torch::rand({1, 3, 224, 224}, torch::dtype(torch::kFloat32).device(torch::kCUDA));

  // run
  auto module = torch::jit::load(argv[1], torch::kCUDA);
  assert(module != nullptr);
  std::cout << "ok" << std::endl;

  std::vector<torch::jit::IValue> inputs;
  inputs.push_back(tensor);

  auto out = module->forward(inputs).toTensor();
  std::cout << out << std::endl;
  std::cout << "done" << std::endl;
}
```

When I run the C++ program, here's the output I receive:

```
(base) [build]$ rm -rf *; cmake -DCMAKE_PREFIX_PATH=/home/ec2-user/libtorch/ ..; make -j; ./example-app ../hard.pt
-- The C compiler identification is GNU 5.3.1
-- The CXX compiler identification is GNU 5.3.1
-- Check for working C compiler: /opt/rh/devtoolset-4/root/usr/bin/cc
-- Check for working C compiler: /opt/rh/devtoolset-4/root/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/rh/devtoolset-4/root/usr/bin/c++
-- Check for working CXX compiler: /opt/rh/devtoolset-4/root/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda (found version "8.0")
-- Caffe2: CUDA detected: 8.0
-- Caffe2: CUDA nvcc is: /usr/local/cuda/bin/nvcc
-- Caffe2: CUDA toolkit directory: /usr/local/cuda
-- Caffe2: Header version is: 8.0
-- Found CUDNN: /usr/local/cuda/include
-- Found cuDNN: v7.1.3  (include: /usr/local/cuda/include, library: /usr/local/cuda/lib64/libcudnn.so)
-- Autodetected CUDA architecture(s):  3.7 3.7 3.7 3.7 3.7 3.7 3.7 3.7
-- Added CUDA NVCC flags for: -gencode;arch=compute_37,code=sm_37
-- Found torch: /home/ec2-user/libtorch/lib/libtorch.so
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ec2-user/example-app/build
Scanning dependencies of target example-app
[ 50%] Building CXX object CMakeFiles/example-app.dir/example-app.cpp.o
[100%] Linking CXX executable example-app
[100%] Built target example-app
ok
Columns 1 to 10-0.1336 -0.7583 -0.0227  0.0581  0.0842 -0.2745  0.5487  1.0180  0.0602  0.0120
...
...
...
Columns 991 to 1000-0.6911  1.0201  0.2691  0.2674 -0.4418  0.1372  0.3234  0.1259 -0.3892  0.1433
[ Variable[CUDAType]{1,1000} ]
done
Segmentation fault
```

## Things I've tried
- I've tried running the above C++ code on the CPU (so, replacing `kCUDA` with `kCPU`). This doesn't result in a segmentation fault.

- I've tried running the above Python/C++ code on both CPU, GPU for a simple feedforward net (30 inputs -> 20 hiddens -> ReLU -> 2 outputs -> Sigmoid); this doesn't result in a segmentation fault. Running with `valgrind` on CPU, I don't see any issues. On GPU, however here's what I get:

```==60061== Memcheck, a memory error detector
==60061== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==60061== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==60061== Command: ./example-app ../simple.pt
==60061==
==60061== Warning: set address range perms: large range [0x395de000, 0x66b46000) (defined)
==60061== Warning: noted but unhandled ioctl 0x30000001 with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: noted but unhandled ioctl 0x27 with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: noted but unhandled ioctl 0x7ff with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: noted but unhandled ioctl 0x25 with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: noted but unhandled ioctl 0x17 with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: set address range perms: large range [0x1000000000, 0x8a00000000) (noaccess)
==60061== Warning: set address range perms: large range [0x8a00000000, 0xa200000000) (noaccess)
==60061== Warning: noted but unhandled ioctl 0x19 with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: noted but unhandled ioctl 0x21 with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==60061== Warning: noted but unhandled ioctl 0x1b with no size/direction hints.
==60061==    This could cause spurious value errors to appear.
==60061==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
ok
 0.4127  0.4883
[ Variable[CUDAType]{1,2} ]
done
==60061==
==60061== HEAP SUMMARY:
==60061==     in use at exit: 1,247,853,758 bytes in 1,312,577 blocks
==60061==   total heap usage: 2,763,800 allocs, 1,451,223 frees, 1,808,988,116 bytes allocated
==60061==
==60061== LEAK SUMMARY:
==60061==    definitely lost: 440 bytes in 7 blocks
==60061==    indirectly lost: 2,344 bytes in 24 blocks
==60061==      possibly lost: 403,300,908 bytes in 155,594 blocks
==60061==    still reachable: 844,550,066 bytes in 1,156,952 blocks
==60061==                       of which reachable via heuristic:
==60061==                         stdstring          : 438,097 bytes in 6,137 blocks
==60061==         suppressed: 0 bytes in 0 blocks
==60061== Rerun with --leak-check=full to see details of leaked memory
==60061==
==60061== For counts of detected and suppressed errors, rerun with: -v
==60061== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)```

## Expected behavior

In running with hard.pt, I expect the code to run without any segfaults.

## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0.dev20190425
 - OS (e.g., Linux): Red Hat Enterprise Linux Server 7.4 (Maipo)
 - How you installed PyTorch (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): `N/A`
 - Python version: `3.7`
 - CUDA/cuDNN version: 8.0.44/7
 - GPU models and configuration: ```GPU models and configuration:
GPU 0: Tesla K80
GPU 1: Tesla K80
GPU 2: Tesla K80
GPU 3: Tesla K80
GPU 4: Tesla K80
GPU 5: Tesla K80
GPU 6: Tesla K80
GPU 7: Tesla K80
```

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] numpydoc==0.8.0
[pip] torch==1.0.1.post2
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.1                      144
[conda] mkl-service               1.1.2            py37he904b0f_5
[conda] mkl_fft                   1.0.6            py37hd81dba3_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] pytorch                   1.0.1           py3.7_cuda8.0.61_cudnn7.1.2_2    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch

## Additional context

I've tried following the instructions to resolve this issue here: https://github.com/pytorch/pytorch/issues/12705 but haven't seen success; I've tried building pytorch from source + linking my C++ code to the generated libs, but this doesn't work either.

cc @ezyang @gchanan @zou3519 @ngimel

**Comment by SpringSJTU** [comment]
The same bug occur to me, do you solve it?

**Comment by karthiksn** [comment]
@SpringSJTU - haven't solved it yet, but haven't actively been working on this.

**Comment by liuchang8am** [comment]
same bug here,

`torch::Tensor out_tensor = module->forward(inputs).toTensor();`

the out_tensor type is [ Variable[CUDAType]{1,10} ], and Segmentation Fault is followed.

**Comment by hshustc** [comment]
The same bug. so sad.

**Comment by li-ren** [comment]
The same problem. QAQ 
Does anyone have a solution?

**Comment by siddhantk** [comment]
Encountered the same bug. Anyone solved it yet?

**Comment by li-ren** [comment]
I solved it. Use pytorch nightly build 🎉

**Comment by vusraju** [comment]
Add "cudaDeviceSynchronize();" before and after you do inference. it solved my issue. Also It didn't change the speed.

**Comment by walton-wang929** [comment]
@vusraju can you paste your codes here?

**Comment by vusraju** [comment]
can't post it. But, just try to synchronize before and after your function/code that deals with gpu computations (forward in this case) and see if it works.

**Comment by buptlj** [comment]
[ Variable[CUDAType]{1,5} ]
Segmentation fault (core dumped)

**Comment by yaoqi-zd** [comment]
same problem, sad

**Comment by powerlic** [comment]
for me, the fault is caused from the cmake file dependencies.  I complied the opencv with cuda-on option and add this to my project.
following cmake code cause error:
set(DEP_LIBS ${CUDA_LIBRARIES}  ${TORCH_LIBRARIES} ${OpenCV_LIBS})
changed this to:
set(DEP_LIBS ${TORCH_LIBRARIES}  ${CUDA_LIBRARIES} ${OpenCV_LIBS})
is ok.
may this help you!

**Comment by deeplearning666** [comment]
@karthiksn Have you solved the problems? I met the same question with you! I'm looking for ward to hearing from you soon!

**Comment by karthiksn** [comment]
Haven't solved this, but also haven't worked with this in over a year :)

**Comment by ezyang** [comment]
This is still affecting 1.5 and it appears to be some sort of problem when you link against cudart too many times. I am not sure if there are downstream consequences of doing this, but if you change your cmake line to

```
target_link_libraries(example-app torch)
```

(instead of `$TORCH_LIBRARIES`) that seems to reduce the number of times we load cudart enough so that the problem goes away.

Also, weirdly, if you run your example-app as `./example-app` inside the directory it was built in there is no segfault, but if you run it in a different directory `build/example-app` it segfaults. I haven't checked if this is a CUDA 9 specific problem.

**Comment by ezyang** [comment]
Here is an ldd from an executable that doesn't segfault:

```
linux-vdso.so.1 =>  (0x00007fffbe1f0000)
libtorch.so => /home/ezyang/local/test/libtorch/lib/libtorch.so (0x00007f8dd5eb2000)
libtorch_cpu.so => /home/ezyang/local/test/libtorch/lib/libtorch_cpu.so (0x00007f8dc7116000)
libtorch_cuda.so => /home/ezyang/local/test/libtorch/lib/libtorch_cuda.so (0x00007f8d97dee000)
libc10.so => /home/ezyang/local/test/libtorch/lib/libc10.so (0x00007f8d97b94000)
libcudart.so.9.2 => /usr/local/cuda-9.2/lib64/libcudart.so.9.2 (0x00007f8d9792a000)
libnvToolsExt.so.1 => /usr/local/cuda-9.2/lib64/libnvToolsExt.so.1 (0x00007f8d97721000)
libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f8d9741a000)
libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f8d97204000)
libc.so.6 => /lib64/libc.so.6 (0x00007f8d96e36000)
libgomp-7c85b1e2.so.1 => /home/ezyang/local/test/libtorch/lib/libgomp-7c85b1e2.so.1 (0x00007f8d96c0c000)
libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f8d969f0000)
librt.so.1 => /lib64/librt.so.1 (0x00007f8d967e8000)
libdl.so.2 => /lib64/libdl.so.2 (0x00007f8d965e4000)
libm.so.6 => /lib64/libm.so.6 (0x00007f8d962e2000)
libcudart-72ec04ea.so.9.2 => /home/ezyang/local/test/libtorch/lib/libcudart-72ec04ea.so.9.2 (0x00007f8d96075000)
/lib64/ld-linux-x86-64.so.2 (0x00007f8dd60c6000)
libc10_cuda.so => /home/ezyang/local/test/libtorch/lib/libc10_cuda.so (0x00007f8d95e46000)
libnvToolsExt-3965bdd0.so.1 => /home/ezyang/local/test/libtorch/lib/libnvToolsExt-3965bdd0.so.1 (0x00007f8d95c3c000)
```

and an ldd from an executable that does:

```
linux-vdso.so.1 =>  (0x00007fff72318000)
libtorch.so => /home/ezyang/local/test/libtorch/lib/libtorch.so (0x00007fe6f3670000)
libc10_cuda.so => /home/ezyang/local/test/libtorch/lib/libc10_cuda.so (0x00007fe6f3441000)
libc10.so => /home/ezyang/local/test/libtorch/lib/libc10.so (0x00007fe6f31e7000)
libnvrtc.so.9.2 => /usr/local/cuda-9.2/lib64/libnvrtc.so.9.2 (0x00007fe6f1bdd000)
libnvToolsExt.so.1 => /usr/local/cuda-9.2/lib64/libnvToolsExt.so.1 (0x00007fe6f19d4000)
libcudart.so.9.2 => /usr/local/cuda-9.2/lib64/libcudart.so.9.2 (0x00007fe6f176a000)
libtorch_cpu.so => /home/ezyang/local/test/libtorch/lib/libtorch_cpu.so (0x00007fe6e29ce000)
libtorch_cuda.so => /home/ezyang/local/test/libtorch/lib/libtorch_cuda.so (0x00007fe6b36a6000)
libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fe6b339f000)
libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fe6b3189000)
libc.so.6 => /lib64/libc.so.6 (0x00007fe6b2dbb000)
libcudart-72ec04ea.so.9.2 => /home/ezyang/local/test/libtorch/lib/libcudart-72ec04ea.so.9.2 (0x00007fe6b2b4e000)
libm.so.6 => /lib64/libm.so.6 (0x00007fe6b284c000)
libgomp-7c85b1e2.so.1 => /home/ezyang/local/test/libtorch/lib/libgomp-7c85b1e2.so.1 (0x00007fe6b2622000)
libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fe6b2406000)
/lib64/ld-linux-x86-64.so.2 (0x00007fe6f3884000)
libdl.so.2 => /lib64/libdl.so.2 (0x00007fe6b2202000)
librt.so.1 => /lib64/librt.so.1 (0x00007fe6b1ffa000)
libnvToolsExt-3965bdd0.so.1 => /home/ezyang/local/test/libtorch/lib/libnvToolsExt-3965bdd0.so.1 (0x00007fe6b1df0000)
```

**Comment by TheBobbyliu** [comment]
I'm using LibTorch 1.4. 
After changing the linking order of CUDA and LibTorch libraries in my own makefile, it seems to work. Here's my solution.
this will cause segmentation fault when leaving:
`
# cuda
INCLUDE_DIRS    += -I /usr/local/cuda/include/
EXTERNAL_LIBS   += -L/usr/local/cuda/lib64 -lcuda -lcudart -lcublas -lcurand -lcudnn -Wl,-rpath,/usr/local/cuda/lib64

# torch
INCLUDE_DIRS    += -I ./third_party/torch/include
INCLUDE_DIRS    += -I ./third_party/torch/include/torch/csrc/api/include
EXTERNAL_LIBS   += -L./lib/torch/ -ltorch -lc10 -lc10_cuda -Wl,-rpath,./lib/torch/
`
this wouldn't:
`
# libtorch
INCLUDE_DIRS    += -I ./third_party/torch/include
INCLUDE_DIRS    += -I ./third_party/torch/include/torch/csrc/api/include
EXTERNAL_LIBS   += -L./lib/torch/ -ltorch -lc10 -lc10_cuda -Wl,-rpath,./lib/torch/
# cuda
INCLUDE_DIRS    += -I /usr/local/cuda/include/
EXTERNAL_LIBS   += -L/usr/local/cuda/lib64 -lcuda -lcudart -lcublas -lcurand -lcudnn -Wl,-rpath,/usr/local/cuda/lib64
`
---
I've checked core file after the segmentation fault, and I find that this fault is happened in cudnn. maybe there's some implicit confliction between libtorch and cudnn?

**Comment by padmasreenagarajan** [comment]
Even i got the same error for my YOLOV3 model but then after 2 full days of learning about seg faults and debugging using gdb debugger in linux ,the seg fault got resolved when i built using Libtorch nightly version.
Previously i got seg fault for Libtorch 1.5 with cu 10.1
Do try and let me know for the updates.,
@TheBobbyliu @ezyang @sflc6

**Comment by Jelly123456** [comment]
@padmasreenagarajan 
Can you give more details on how do you solve the error "segmentation fault(core dumped)"? 
I am facing the same problem now. Your inputs will be very helpful.

Environment details:
    OS: SUSE Linux Enterprise Server 12 SP3
    Libtorch: Preview(Nightly) 
    CUDA: V10.2
    cuDNN:  v8.0.3

**Comment by padmasreenagarajan** [comment]
Hello @Jelly123456 ,

    This is a bug associated with C++ API-LibTorch (Stable) -1.5.1 when destructing cudnn handle pool (in CUDA-10.1)

**SOLUTION:**
                      This is solved in LibTorch Preview (Nightly) Build.

Since Nightly build is a research package and a stable version of PyTorch 1.6 was released that time, I have generated the build files with LibTorch 1.6 stable and executed all the test images as shown below without any errors (Segmentation fault is resolved)

So, I would suggest you to install the latest PyTorch version and the corresponding LibTorch stable version.
If it still doesn't work, try installing CUDA 10.1 and let me know your issues if any!

Regards,
Padmasree N.

**Comment by malfet** [comment]
@Jelly123456 Can you preproduce the crash using sample in the issue description? Or something more involved?

**Comment by ngimel** [comment]
Closing due to age.
